feat: build graph and find fastest path

JackSMunn opened the pull request:
This pull request includes significant changes to the codebase, focusing on adding new functionalities for data fetching, graph building, and pathfinding using sensor and room data. The changes also include updates to the main application logic and some utility functions.

### New functionalities:

* [`app/classes/data_fetcher.py`](diffhunk://#diff-7fa766977f7b1eacdf7b78f22d721a6f63c8540459b1f6275a7e7413a9b9f438R1-R32): Introduced `DataFetcher` class to fetch sensor and room data from an API and convert them into respective objects.
* [`app/classes/sensor_graph.py`](diffhunk://#diff-ceb90e18cc33f771b88b66d13c001e4d8bad274677cf60c8ff722ca66c508fb9R1-R57): Added `SensorGraph` class to build a graph of sensors and find the fastest path using Dijkstra's algorithm.

### New classes:

* [`app/classes/room.py`](diffhunk://#diff-781a99729ac6aeb78e3ebcf5c734d0720055e1c07aebcc5dbc8e05feec254225R1-R14): Added `Room` class to represent room data with a method to create an instance from a dictionary.
* [`app/classes/sensor.py`](diffhunk://#diff-98906d07c09411ab47b33866ee6a26f6e20d0e030db2bc738e8313910182f3daR1-R18): Added `Sensor` class to represent sensor data with methods to create an instance from a dictionary and attach room objects.

### Main application updates:

* [`app/main.py`](diffhunk://#diff-990f7e4140fab1ad82afc787505090b64d4024e63a891405cd9085e7e6b249ddL2-R6): Updated the main application to initialize data fetcher, retrieve sensors and rooms, build a sensor graph, and find the fastest path between two sensors. [[1]](diffhunk://#diff-990f7e4140fab1ad82afc787505090b64d4024e63a891405cd9085e7e6b249ddL2-R6) [[2]](diffhunk://#diff-990f7e4140fab1ad82afc787505090b64d4024e63a891405cd9085e7e6b249ddR20-R73)

### Utility functions:

* [`app/classes/utils.py`](diffhunk://#diff-b7a804bcb4fd9686b26bc7497a76f88ac280a132ba29ab0504ea9192c2513296R1-R12): Added a utility function `dict_to_obj` to convert dictionaries to objects for attribute-style access.

### Configuration updates:

* [`.coveragerc`](diffhunk://#diff-834e9b406d74791ffbafaeec9cc894082cea9739bc347b6ff9f72312c92ebc79R1-R11): Added a new coverage configuration file to specify which files to omit from coverage and set reporting options.

JackSMunn committed:
test: changed get_room.py

- changed get_room.py to allow for the correct way of using the database

JackSMunn committed:
feat: populating database, draw graph

- added draw_sensor_query.py that draws graph based on rooms and sensors from the DB
- added populate.py, that populates database with "dummy" data, but real graph structure

JackSMunn committed:
update: requirements.txt

JackSMunn committed:
feat: draw graph, use dijkstra for fastest path

- changed draw_sensor_query.py, to allow for drawing the graph with the fastest path
- changed draw_sensor_query.py to use dijkstra to find fastest path in graph
- added rest of the rooms on floor 2 of SMK floorplan based on open.SMK image (from the API)

JackSMunn committed:
feat: graph through endpoint

- added the modified version of 1st floor and 2nd floor of SMK, as it is beneficial to have if any changes are to be made.
- update: moved route_model.py and route_service.py to dir: app
- added database_draw_graph.py, that directly communicates with database to draw the graph
- added endpoint_draw_graph.py  that communicates with sensorsim/dataprocessing repo to draw the graph
- deleted populate.py (moved to database repo)
- update: requirements to include "requests"

JackSMunn committed:
feat: split graph.py into classes + unittesting

- added cov file for reporting on coverage
- added init files to ensure package recognising
- update: split graph into classes to follow OOP guidelines:
- added data_fetcher.py that fetches data from url
- added room and sensor classes that matches model (room only using crowdfactor currently, sensor only using rooms)
- added sensor_graph.py that builds graph and finds fastest path using dijkstra
- added utils.py that takes dictionary and turns into object
- updated main.py to run the functions from the other classes, does nolonger expose endpoint as of now.
- added not_used files database_draw_graph.py, endpoint_draw_graph.py, get_room.py, graph.py, these files are not to be used, but remain as a reminder of past transgressions of immorality
- added test_data_fetcher.py
- added test_room.py
- added test_sensor.py
- added test_sensor_graph.py
- added test_utils.py
- added pytest.ini file to configure pytest
- updated requirements file to include pytest-cov

JackSMunn committed:
feat: added linting and test to github actions

- added lint.yml
- added test.yml

Mafusn commented on app/classes/sensor_graph.py:
build_graph() should be split into more functions so we don't have 3 nested for loops with double nested if statements

JackSMunn replied:
Should be better now.

Mafusn reviewed (changes requested):
Needs some cleanup of the build_graph() as mentioned below

JackSMunn committed:
Merge branch 'dev' into db-query-test

JackSMunn committed:
fix: pytest.ini

JackSMunn committed:
Merge branch 'db-query-test' of https://github.com/SW8-Indoor-Crowded-Regions/pathfinding into db-query-test

JackSMunn committed:
update: coverage excludes

JackSMunn committed:
update: requirements.txt

JackSMunn committed:
fix: linting, removed some imports

- added ruff.toml file that excludes some files, and sets linting style

JackSMunn committed:
feat: graph now multigraph and load/save functions

- changed sensor_graph.py to allow for multigraph, but collapses if multi found
- saves and loads a graph from a picklefile --> to minimize resources needed

JackSMunn committed:
update: changed from multigraph to singleEgraph

- changed how edges work (from multi, to single, always)
- changed tests accordingly
- changed tests to properly represent how the sensors work

Mafusn reviewed (approved):
LGTM 💅

MortenMunk reviewed (approved):
LGTM - I did not look too much into /not_used dir, as instructed by Mr. Skaarup
